src/flist/mount.rs

```diff
@@ -0,0 +1,487 @@
+    pub flist: PathBuf,
+    pub cache: PathBuf,
+    pub mountpoint: PathBuf,
+    pub ro: PathBuf,
+    pub log: PathBuf,
+    pub syscalls: S,
+    pub storage: A,
+    pub executor: E,
+    pub db: MetadataDbMgr,
+}
+impl<A, S, E> MountManager<A, S, E>
+where
+    A: VolumeAllocator + Sync + Send,
+    S: Syscalls + Sync + Send,
+    E: Executor + Sync + Send,
+{
+    pub async fn new<R: Into<PathBuf>>(
+        root: R,
+        syscalls: S,
+        storage: A,
+        executor: E,
+    ) -> Result<Self>
+    where
+        R: AsRef<str>,
+    {
+        let root = root.into();
+        let db = MetadataDbMgr::new(root.join("flist")).await?;
+        fs::create_dir_all(&root).await?;
+        // prepare directory layout for the module
+        for path in &["flist", "cache", "mountpoint", "ro", "log"] {
+            fs::create_dir_all(&root.join(path)).await?;
+        }
+        Ok(Self {
+            flist: root.join("flist"),
+            cache: root.join("cache"),
+            mountpoint: root.join("mountpoint"),
+            ro: root.join("ro"),
+            log: root.join("log"),
+            root,
+            syscalls,
+            storage,
+            executor,
+            db,
+        })
+    }
+
+    // returns the mount path out of an flist name simplly joins /<FLISTS_ROOT>/<mountpoint>/<name>
+    // this where this flist instance will be
+    pub fn mountpath<T: AsRef<str>>(&self, name: T) -> Result<PathBuf> {
+        let mountpath = self.mountpoint.join(name.as_ref());
+        if mountpath.parent() != Some(self.mountpoint.as_path()) {
+            bail!("inavlid mount name: {}", name.as_ref());
+        }
+        Ok(mountpath)
+    }
+
+    // returns ro path joined with flist hash
+    // this where we mount the flist for read only
+    fn flist_ro_mount_path<R: AsRef<str>>(&self, hash: R) -> Result<PathBuf> {
+        let mountpath = self.ro.join(hash.as_ref());
+        if mountpath.parent() != Some(self.ro.as_path()) {
+            bail!("invalid mount name")
+        }
+
+        Ok(mountpath)
+    }
+
+    // Checks if the given path is mountpoint or not
+    pub async fn is_mounted<P: AsRef<Path>>(&self, path: P) -> bool {
+        matches!(storage::mountpoint(path.as_ref()).await, Ok(Some(_)))
+    }
+
+    // Checks is the given path is a valid mountpoint means:
+    // it is either doesn't exist or
+    // it is a dir and not a mountpoint for anything
+    pub async fn valid<P: AsRef<Path>>(&self, path: P) -> bool {
+        match fs::metadata(&path).await {
+            Ok(info) => info.is_dir() && !self.is_mounted(&path).await,
+            Err(err) if err.kind() == io::ErrorKind::NotFound => true,
+            Err(err) if err.kind() == io::ErrorKind::ConnectionAborted => {
+                matches!(self.syscalls.umount(path, None), Ok(_))
+            }
+            _ => false,
+        }
+    }
+
+    async fn wait_mountpoint<P: AsRef<Path>>(&self, path: P, seconds: u32) -> Result<()> {
```

Comment:
This function will always fail if the `seconds == 0` it needs to check at least once and wait if need to before it tries again.